File: packages/runtime-go/adapters/pubsub.go
package adapters

import (
	"context"
	"encoding/json"
	"fmt"

	"cloud.google.com/go/pubsub"
	runtime "github.com/eventgen/runtime-go"
)

type PubSubTransportAdapter struct {
	topic *pubsub.Topic
}

func NewPubSubTransportAdapter(ctx context.Context, projectID, topicID string) (*PubSubTransportAdapter, error) {
	client, err := pubsub.NewClient(ctx, projectID)
	if err != nil {
		return nil, fmt.Errorf("pubsub adapter: new client: %w", err)
	}
	return &PubSubTransportAdapter{topic: client.Topic(topicID)}, nil
}

func NewPubSubTransportAdapterWithTopic(topic *pubsub.Topic) *PubSubTransportAdapter {
	return &PubSubTransportAdapter{topic: topic}
}

func (a *PubSubTransportAdapter) Publish(ctx context.Context, envelope runtime.EventEnvelope) (*runtime.PublishResult, error) {
	data, err := json.Marshal(envelope)
	if err != nil {
		return nil, fmt.Errorf("pubsub adapter: marshal envelope: %w", err)
	}
	result := a.topic.Publish(ctx, &pubsub.Message{
		Data: data,
		Attributes: map[string]string{
			"eventId": envelope.EventID,
			"version": envelope.Version,
			"traceId": envelope.Metadata.TraceID,
		},
	})
	id, err := result.Get(ctx)
	if err != nil {
		return nil, fmt.Errorf("pubsub adapter: publish %s: %w", envelope.EventID, err)
	}
	return &runtime.PublishResult{MessageID: id}, nil
}

func (a *PubSubTransportAdapter) Stop() {
	a.topic.Stop()
}

File: event-sdk-generator/packages/runtime-typescript/src/adapters/pubsub.ts
import { PubSub } from "@google-cloud/pubsub";
import type { TransportAdapter, EventEnvelope, PublishResult } from "../types.js";

export type PubSubTransportAdapterOptions = {
  topicName: string;
  projectId?: string;
  /** Pass an existing PubSub instance to share credentials */
  pubsub?: PubSub;
};

export class PubSubTransportAdapter implements TransportAdapter {
  private readonly pubsub: PubSub;
  private readonly topicName: string;

  constructor(options: PubSubTransportAdapterOptions) {
    this.topicName = options.topicName;
    this.pubsub = options.pubsub ?? new PubSub({ projectId: options.projectId });
  }

  async publish(envelope: EventEnvelope): Promise<PublishResult> {
    const data = Buffer.from(JSON.stringify(envelope));
    const messageId = await this.pubsub.topic(this.topicName).publishMessage({
      data,
      attributes: {
        eventId: envelope.eventId,
        version: envelope.version,
        traceId: envelope.metadata.traceId,
      },
    });
    return { messageId };
  }
}
